tls/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+        self.verifier.verify(&pki_path).map_err(|e| {
+            error!(self.log, "Failed to verify cert: {e}");
+            rustls::Error::InvalidCertificate(
+                rustls::CertificateError::BadEncoding,
+            )
+        })?;
+
+        info!(self.log, "Certificate chain verified successfully");
+
+        Ok(())
+    }
+
+    fn verify_signature(
+        &self,
+        message: &[u8],
+        cert: &rustls::pki_types::CertificateDer<'_>,
+        sig: &[u8],
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, rustls::Error>
+    {
+        // Get the public key
+        let cert = Certificate::from_der(cert).map_err(|_| {
+            rustls::Error::InvalidCertificate(
+                rustls::CertificateError::BadEncoding,
+            )
+        })?;
+
+        let pubkey: [u8; 32] = match cert
+            .tbs_certificate
+            .subject_public_key_info
+            .subject_public_key
+            .as_bytes()
+        {
+            Some(pubkey) => pubkey.try_into().map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?,
+            None => {
+                return Err(rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                ))
+            }
+        };
+        let verifying_key = ed25519_dalek::VerifyingKey::from_bytes(&pubkey)
+            .map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?;
+
+        // We must hash with SHA-512 and then verify the digest
+        let mut prehashed = Sha512::new();
+        prehashed.update(message);
+
+        let signature =
+            ed25519_dalek::Signature::from_slice(sig).map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadEncoding,
+                )
+            })?;
+        verifying_key
+            .verify_prehashed(prehashed, Some(TLS_SIGNING_CONTEXT), &signature)
+            .map_err(|_| {
+                rustls::Error::InvalidCertificate(
+                    rustls::CertificateError::BadSignature,
+                )
+            })?;
+
+        info!(self.log, "Signature verified successfully");
+
+        Ok(HandshakeSignatureValid::assertion())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rustls::client::ResolvesClientCert;
+    use rustls::server::ResolvesServerCert;
+    use slog::Drain;
+    use std::{
+        io::ErrorKind,
+        net::{TcpListener, TcpStream},
+    };
+
+    fn logger() -> slog::Logger {
+        let decorator = slog_term::TermDecorator::new().build();
+        let drain = slog_term::FullFormat::new(decorator).build().fuse();
+        let drain = slog_async::Async::new(drain).build().fuse();
```

Comment:
hmm, do we _really_ want to use `slog_async` for logging in tests? i think that in tests, my preference would be that logs never get discarded if a buffer fills up, even if it means a higher performance cost. we don't *really* care to avoid blocking the tests on writing some logs nearly as much as we would care about not losing some of those logs, which we might debug a potential test failure...

Reply:
Yeah, I guess no reason to use it. This was just me doing some copypasta.